dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -441,65 +443,85 @@ Segment::SegmentMetaInfos Segment::readAllSegmentsMetaInfoInRange( //
     PageIdU64 current_segment_id = DELTA_MERGE_FIRST_SEGMENT_ID;
     auto end_to_segment_id_cache = checkpoint_info->checkpoint_data_holder->getEndToSegmentIdCache(
         KeyspaceTableID{context.keyspace_id, context.physical_table_id});
-    auto lock = end_to_segment_id_cache->lock();
-    bool is_cache_ready = end_to_segment_id_cache->isReady(lock);
-    if (is_cache_ready)
+    bool is_cache_ready = false;
     {
-        current_segment_id
-            = end_to_segment_id_cache->getSegmentIdContainingKey(lock, target_range.getStart().toRowKeyValue());
+        auto lock = end_to_segment_id_cache->readLock();
+        is_cache_ready = end_to_segment_id_cache->isReady(lock);
     }
-    LOG_DEBUG(Logger::get(), "Read segment meta info from segment {}", current_segment_id);
-    std::vector<std::pair<DM::RowKeyValue, UInt64>> end_key_and_segment_ids;
-    SegmentMetaInfos segment_infos;
-    while (current_segment_id != 0)
-    {
-        Segment::SegmentMetaInfo segment_info;
-        auto target_id = UniversalPageIdFormat::toFullPageId(
-            UniversalPageIdFormat::toFullPrefix(context.keyspace_id, StorageType::Meta, context.physical_table_id),
-            current_segment_id);
-        auto page = checkpoint_info->temp_ps->read(target_id, nullptr, {}, false);
-        if unlikely (!page.isValid())
-        {
-            // After #7642, DELTA_MERGE_FIRST_SEGMENT_ID may not exist, however, such checkpoint won't be selected.
-            // If it were to be selected, the FAP task could fallback to regular snapshot.
-            throw Exception(
-                ErrorCodes::LOGICAL_ERROR,
-                "Can't find page id {}, keyspace={} table_id={} current_segment_id={} range={}",
-                target_id,
-                context.keyspace_id,
-                context.physical_table_id,
-                current_segment_id,
-                target_range.toDebugString());
-        }
-        segment_info.segment_id = current_segment_id;
-        ReadBufferFromMemory buf(page.data.begin(), page.data.size());
-        readSegmentMetaInfo(buf, segment_info);
-        if (!is_cache_ready)
+
+    auto protected_logic = [&](GenericLock && lock) {
+        if (is_cache_ready)
         {
-            end_key_and_segment_ids.emplace_back(segment_info.range.getEnd().toRowKeyValue(), segment_info.segment_id);
+            current_segment_id = end_to_segment_id_cache->getSegmentIdContainingKey(
+                std::get<std::shared_lock<std::shared_mutex>>(lock),
+                target_range.getStart().toRowKeyValue());
         }
-        current_segment_id = segment_info.next_segment_id;
-        if (!(segment_info.range.shrink(target_range).none()))
+        LOG_DEBUG(Logger::get(), "Read segment meta info from segment {}", current_segment_id);
```

Comment:
Generate a logger with the fap related information as suffix for filtering these logging? For example, region_id

Reply:
IMO, A segment is not necessarily bound to a region. Actually, region_id is already stored in checkpoint_info, so I just use it directly.

Reply:
When doing troubleshooting, many logs are print into the log files. Adding region_id to the logging suffix can help us quickly filter what happens when the FAP of given region_id.

Reply:
I have already attached the region_id into the printed log.
What I said is that we directly use the region_id in checkpoint_info, rather than compose a new Logger

Reply:
I have changed `log` in FastAddPeerWrite and readAllSegmentsMetaInfoInRange. For FastAddPeerSelect, I will changed it in another PR after the pick of compation.